### PR TITLE
Add option to remove empty strings in YAML

### PR DESCRIPTION
**Pull request by dbendilas**
**TODO**
- [x] Remove comments of deleted entries
- [x] Optimize template rebuilding to reduce memory footprint

Problem and/or solution
-----------------------
We need to be able to remove entries with no translation but at the same time use the template.

The previous implementation, when instructed to only use the entries that had a translation, didn't use the template at all. Instead, it generated a new document from scratch, thus losing the initial structure, comments etc.

With the approach this PR is introducing, there is no need to ditch the template anymore in such cases, so the resulting document has the same structure as before.

How to test
-----------
Unit tests.

Manual testing through Transifex (related PRs pending)
- upload YAML resource
- translate some strings only
- download for use
- make sure that:
  - is valid YAML
  - untranslated strings are not present in the file
  - the overall structure is correct
    - only empty strings and their corresponding comments are missing
    - parent dictionaries of removed strings are not deleted
  - anything else that I may have missed!

Reviewer checklist
------------------

Code:
* [ ] Change is covered by unit-tests
* [ ] Code is well documented, well styled and is following [best practices](https://tem.transifex.com)
* [ ] Performance issues have been taken under consideration
* [ ] Errors and other edge-cases are handled properly

PR:
* [ ] Problem and/or solution are well-explained
* [ ] Commits have been squashed so that each one has a clear purpose
* [ ] Commits have a proper commit message [according to TEM](https://tem.transifex.com/github-guide.html#working-on-a-feature)

### CODE REVIEW

File: openformats/formats/yaml/yaml.py [hunk 1]
```diff
@@ -125,10 +123,56 @@ def compile(self, template, stringset, **kwargs):
...
+
+        transcriber.copy_until(len(template))
+        compiled = transcriber.get_destination()
+
+        # Remove all marked lines
+        if translations_removed:
+            compiled = '\n'.join(
+                [x for x in compiled.splitlines()
+                 if REMOVE_LINE_INDICATOR not in x]
+            )
```
Comment (dbendilas) [inline openformats/formats/yaml/yaml.py:173]: TODO: I'm going to optimize this, as now it will use a lot of memory in big files. (planning to use `StringIO` and read line by line).

File: openformats/formats/yaml/yaml.py [hunk 2]
```diff
@@ -125,10 +128,59 @@ def compile(self, template, stringset, **kwargs):
...
+            if not translation:
+                transcriber.add(REMOVE_LINE_INDICATOR)
+                translations_removed = True
+
+                # If folded style is used, we need to add an extra newline,
+                # otherwise the next entry will be on the same line as
+                # the indicator, and will also be removed
+                style = string.flags.split(':')[-1]
+                if style == '>':
+                    transcriber.add('\n')
```
Comment (dbendilas) [inline openformats/formats/yaml/yaml.py:169]: Note for reviewers: see https://learnxinyminutes.com/docs/yaml/ for more information on folded style.

**Comment by deathbird** [comment]
Remove local testbed db file `testbed/testbed.sqlite` from the commited files.

**Comment by dbendilas** [comment]
> Remove local testbed db file testbed/testbed.sqlite from the commited files.

Oops. Done, good catch!

**Comment by deathbird** [comment]
I did a manual test in YAML generic format at first and though most of it was correct, I noticed some keys, mainly nested or nested lists, that got into the `for_use` export file while they aren't translated.

I used the following source file:
```
---
title: About writing and formatting on GitHub
intro: into text
numeric_var: 12.5

# one comment

key1:
  - list_key:
    - li within li
      # second comment
    - li within li 2
  - li 2
key2:
  - object_within_list: value
  - "text with a #hashtag in it"
  - li 5
  - 3.5
  - true

description: >
  folded
  style
  text
custom_vars:
  var1: "text: some value"
  var2: |
    literal
    style with "quotes"
    text

nested_key_outer:
  nested_key_inner:
    nested_value

key.with.dots: dot value

1: integer key

empty_value: ''

single_quoted_value: 'single quoted value'
value with quote: "'value"
value with start backtick: "`value"
value with end backtick: "value`"

<weird_key>: value
"[weird_key]": value

true: boolean key

double_list:
  - - test

simple_flow_list: ["one", 'two', three]
double_flow_list: [[ test ]]

binary_value: !!binary |
  R0lGODlhDAAMAIQAAP//9/X17unp5WZmZgAAAOfn515eXvPz7Y6OjuDg4J+fn5
  OTk6enp56enmlpaWNjY6Ojo4SEhP/++f/++f/++f/++f/++f/++f/++f/++f/+
  +f/++f/++f/++f/++f/++SH+Dk1hZGUgd2l0aCBHSU1QACwAAAAADAAMAAAFLC
  AgjoEwnuNAFOhpEMTRiggcz4BNJHrv/zCFcLiwMWYNG84BwwEeECcgggoBADs=

# complex key syntax
? - Manchester United
  - Real Madrid
: [ 2001-01-01, 2002-02-02 ]

# anchors and aliases
anchor_key: &an_anchor
  - value
alias_key:
  - *an_anchor
    # comment in nested list
  - "another true value"

# Custom tags
foo:  !test bar # Should treat as string and ignore leading spaces
bar: !xml "foo <xml>bar</xml>" # Also a string
hello: !!str World # Translatable
number: !!int 123 # Should ignore
bin: !!binary aGVsbG8= # Should ignore

# Test with non-ASCII keys
#σχόλιο
σχόλιο: κείμενο
emojis: \uD83D\uDC40 🏹🍍🍍 \U0001F418

anchor_with_label:
  # comment in nested key
  test: something
  anchor_test: &another_anchor     this is another anchor with alias and label
    - value1
    - value2
  testing_alias:
    - *another_anchor
    - *an_anchor
    # comment in nested list
    - "something else"
```
Then I translated in greek some strings as you can see in the following downloaded (for_translate) greek (el) version (that includes all strings translated or not):
```
---
title: 
intro: μές στο κείμενο
numeric_var: 12.5

# one comment

key1:
  - list_key:
    - λίστα μέσα σε λίστα
      # second comment
    - 
  - λίστα 2
key2:
  - object_within_list: 
  - 
  - 
  - 3.5
  - true

description: custom_vars:
  var1: 
  var2: nested_key_outer:
  nested_key_inner:
    

key.with.dots: 

1: 

empty_value: ''

single_quoted_value: 
value with quote: 
value with start backtick: 
value with end backtick: 

<weird_key>: 
"[weird_key]": 

true: 

double_list:
  - - 

simple_flow_list: [, , ]
double_flow_list: [[  ]]

binary_value: !!binary |
  R0lGODlhDAAMAIQAAP//9/X17unp5WZmZgAAAOfn515eXvPz7Y6OjuDg4J+fn5
  OTk6enp56enmlpaWNjY6Ojo4SEhP/++f/++f/++f/++f/++f/++f/++f/++f/+
  +f/++f/++f/++f/++f/++SH+Dk1hZGUgd2l0aCBHSU1QACwAAAAADAAMAAAFLC
  AgjoEwnuNAFOhpEMTRiggcz4BNJHrv/zCFcLiwMWYNG84BwwEeECcgggoBADs=

# complex key syntax
? - Manchester United
  - Real Madrid
: [ 2001-01-01, 2002-02-02 ]

# anchors and aliases
anchor_key: &an_anchor
  - 
alias_key:
  - *an_anchor
    # comment in nested list
  - 

# Custom tags
foo:  !test  # Should treat as string and ignore leading spaces
bar: !xml  # Also a string
hello:  # Translatable
number: !!int 123 # Should ignore
bin: !!binary aGVsbG8= # Should ignore

# Test with non-ASCII keys
#σχόλιο
σχόλιο: 
emojis: \uD83D\uDC40 🏹🍍🍍 \U0001F418

anchor_with_label:
  # comment in nested key
  test: 
  anchor_test: &another_anchor     "Αυτό είναι άλλο ένα άνκορ και μία ετικέτα\n  - τιμή 1\n  - τιμή 2"
  testing_alias:
    - *another_anchor
    - *an_anchor
    # comment in nested list
    - 

```
Finally I downloaded the `for_use` file:
```
---
intro: μές στο κείμενο
numeric_var: 12.5

# one comment

key1:
  - list_key:
    - λίστα μέσα σε λίστα
  - λίστα 2
key2:
  - 3.5
  - true

custom_vars:
  nested_key_inner:



empty_value: ''




double_list:


binary_value: !!binary |
  R0lGODlhDAAMAIQAAP//9/X17unp5WZmZgAAAOfn515eXvPz7Y6OjuDg4J+fn5
  OTk6enp56enmlpaWNjY6Ojo4SEhP/++f/++f/++f/++f/++f/++f/++f/++f/+
  +f/++f/++f/++f/++f/++SH+Dk1hZGUgd2l0aCBHSU1QACwAAAAADAAMAAAFLC
  AgjoEwnuNAFOhpEMTRiggcz4BNJHrv/zCFcLiwMWYNG84BwwEeECcgggoBADs=

# complex key syntax
? - Manchester United
  - Real Madrid
: [ 2001-01-01, 2002-02-02 ]

# anchors and aliases
anchor_key: &an_anchor
alias_key:
  - *an_anchor

number: !!int 123 # Should ignore
bin: !!binary aGVsbG8= # Should ignore

emojis: \uD83D\uDC40 🏹🍍🍍 \U0001F418

anchor_with_label:
  anchor_test: &another_anchor     "Αυτό είναι άλλο ένα άνκορ και μία ετικέτα\n  - τιμή 1\n  - τιμή 2"
  testing_alias:
    - *another_anchor
    - *an_anchor
```
In the `for_use` download the following keys are included in the file while blank:
```
custom_vars:
  nested_key_inner:



empty_value: ''




double_list:
```
I would expect them not be included in the `for_use` file. What do you think?

**Comment by deathbird** [comment]
Also pls update the PR description, as you did in the txc PR, in order to be clear what/why we are doing. As it stands explains only the how (implementation) part.

**Comment by deathbird** [comment]
I did another manual test for the YAML v3 case (ruby) translating the same strings as in my first test and got back more blank keys in the `for_use` download. The input file I used for the `en` language:
```
en:
  normal string: normal string
  pluralized_string:
    one: One ȧɠǿ
    other: Other ȧɠǿ
  another normal string: normal string

  empty_pluralized_string:
    one: ""
    other: ""

  # A correct plural rule
  pluralized_string_with_quotes:
    one: 'this is a test'
    other: "this is a double quote test"

  pluralized_string_with_extra_keys:
    zero: Zero
    one: One
    other: Other
  non_pluralized_string_with_extra_keys:
    extra_key: random value
    one: One
    other: Other
  plural_key_with_nested_nodes:
    one:
      nested_key: One
    other: Other

  nested_key_outer_pluralized:
    # A nested plural rule
    nested_key_inner_pluralized:
      one: 'this is a test one'
      other: "this is a double quote test other"

  # A plural rule with empty strings
  pluralized_string_with_quotes_empty:
    one: ''
    other: ""

  title: About writing and formatting on GitHub
  intro: into text
  numeric_var: 12.5

  # one comment

  key1:
    - list_key:
      - li within li
        # second comment
      - li within li 2
    - li 2
  key2:
    - object_within_list: value
    - "text with a #hashtag in it"
    - li 5
    - 3.5
    - true

  description: >
    folded
    style
    text
  custom_vars:
    var1: "text: some value"
    var2: |
      literal
      style with "quotes"
      text

  nested_key_outer:
    nested_key_inner:
      nested_value

  key.with.dots: dot value

  1: integer key

  empty_value: ''

  single_quoted_value: 'single quoted value'
  value with quote: "'value"
  value with start backtick: "`value"
  value with end backtick: "value`"

  <weird_key>: value
  "[weird_key]": value

  true: boolean key

  double_list:
    - - test

  simple_flow_list: ["one", 'two', three]
  double_flow_list: [[ test ]]

  binary_value: !!binary |
    R0lGODlhDAAMAIQAAP//9/X17unp5WZmZgAAAOfn515eXvPz7Y6OjuDg4J+fn5
    OTk6enp56enmlpaWNjY6Ojo4SEhP/++f/++f/++f/++f/++f/++f/++f/++f/+
    +f/++f/++f/++f/++f/++SH+Dk1hZGUgd2l0aCBHSU1QACwAAAAADAAMAAAFLC
    AgjoEwnuNAFOhpEMTRiggcz4BNJHrv/zCFcLiwMWYNG84BwwEeECcgggoBADs=

  # complex key syntax
  ? - Manchester United
    - Real Madrid
  : [ 2001-01-01, 2002-02-02 ]

  # anchors and aliases
  anchor_key: &an_anchor
    - value
  alias_key:
    - *an_anchor
    # comment in nested list
    - "another true value"

  # Custom tags
  foo:  !test bar # Should treat as string and ignore leading spaces
  bar: !xml "foo <xml>bar</xml>" # Also a string
  hello: !!str World # Translatable
  number: !!int 123 # Should ignore
  bin: !!binary aGVsbG8= # Should ignore

  # Test with non-ASCII keys
  #σχόλιο
  σχόλιο: κείμενο
  emojis: \uD83D\uDC40 🏹🍍🍍 \U0001F418

  anchor_with_label:
    # comment in nested key
    test: something
    anchor_test: &another_anchor     this is another anchor with alias and label
      - value1
      - value2
    testing_alias:
      - *another_anchor
      - *an_anchor
      # comment in nested list
      - "something else"
```
And the `for_use` download is:
```
el:
  pluralized_string:

  empty_pluralized_string:
    one: ""
    other: ""

  # A correct plural rule
  pluralized_string_with_quotes:

  pluralized_string_with_extra_keys:
  non_pluralized_string_with_extra_keys:
  plural_key_with_nested_nodes:
    one:

  nested_key_outer_pluralized:
    # A nested plural rule
    nested_key_inner_pluralized:

  # A plural rule with empty strings
  pluralized_string_with_quotes_empty:
    one: ''
    other: ""

  intro: μές στο κείμενο
  numeric_var: 12.5

  # one comment

  key1:
    - list_key:
      - λίστα μέσα σε λίστα
    - λίστα 2
  key2:
    - 3.5
    - true

  custom_vars:
    nested_key_inner:



  empty_value: ''




  double_list:


  binary_value: !!binary |
    R0lGODlhDAAMAIQAAP//9/X17unp5WZmZgAAAOfn515eXvPz7Y6OjuDg4J+fn5
    OTk6enp56enmlpaWNjY6Ojo4SEhP/++f/++f/++f/++f/++f/++f/++f/++f/+
    +f/++f/++f/++f/++f/++SH+Dk1hZGUgd2l0aCBHSU1QACwAAAAADAAMAAAFLC
    AgjoEwnuNAFOhpEMTRiggcz4BNJHrv/zCFcLiwMWYNG84BwwEeECcgggoBADs=

  # complex key syntax
  ? - Manchester United
    - Real Madrid
  : [ 2001-01-01, 2002-02-02 ]

  # anchors and aliases
  anchor_key: &an_anchor
  alias_key:
    - *an_anchor

  number: !!int 123 # Should ignore
  bin: !!binary aGVsbG8= # Should ignore


  anchor_with_label:
    anchor_test: &another_anchor     Αυτό είναι άλλο ένα άνκορ και μία ετικέτα\n - τιμή 1\n - τιμή 2
    testing_alias:
      - *another_anchor
      - *an_anchor
```
Observe that in addition with the problematic keys of the first test many more blank keys exist in the download that have to do with pluralized keys.

File: openformats/formats/yaml/yaml.py [hunk 3]
```diff
@@ -301,7 +313,7 @@ def _get_key_for_node(self, key, parent_key):
             parent_key += '.' + key
         return parent_key
 
-    def _parse_yaml_data(self, yaml_data, parent_key, context="",
+    def _parse_yaml_data(self, yaml_data, parent_key, parents, context="",
```
Comment (deathbird) [inline openformats/formats/yaml/yaml.py:582]: We want not to include empty keys in YAML download translations `for_use` mode.

For example if we have a source file:
```
# ... keys here ...

key2:
  - object_within_list: value
  - "text with a #hashtag in it"
  - key3:
      key4:
        key5: value5
  - li 5
  - 3.5
  - true

outer:
  inner:
    - text1
    - text2

# ... keys here ...
```

and both `key2.key3.key4.key5` and `inner.outer` don't have `el` translations then, when we download `el` translations 'for_use' we should have:

```
# ... keys here ...

key2:
  - object_within_list: τιμή
  - "κείμενο με χαςταγκ"
  - λίστα 5
  - 3.5
  - true

# ... keys here ...
```

in our db we store:
* openstrings - sourceentities
* template (with hashes)

This commit attempts to store (parsing phase) for every string it's parents in the original content. It does that by the new `parents` field in OpenString class. `parents` is a list of parent node tuples to the current node.
 
Further work should store that info to DB and retrieve it during compiling. Then if the string doesn't have a translation find the farthest parent with empty child nodes (translations) and remove from there. The tricky part is that parent nodes positions should be synced with the template and not the original file since that's what we store in the db and use to produce the compiled file.
